### ISSUE TITLE
Unexpected behaviour when assigning array elements

### ISSUE THREAD

**Issue by MatthieuSchaller** [yt-project/unyt#373]
* unyt version: 2.9.4 (but was always present)
* Python version: 3.10.6 (though not relevant)
* Operating System: irrelevant

### Description

No unit check is performed when assigning elements of an array. This is fairly counter-intuitive behaviour
and can likely lead to user errors downstream.

Here is an example (courtesy in part of @JBorrow):

```python
In [1]: import unyt

In [2]: unyt.__version__
Out[2]: 'v2.9.4'

In [3]: x = unyt.unyt_array([0, 2, 3], "cm")

In [4]: y = unyt.unyt_array([5, 6, 7], "kg")

In [5]: x[1] = 4

In [6]: x[2] = y[0]

In [7]: x
Out[7]: unyt_array([0, 4, 5], 'cm')
```

Lines 5 and 6 show two different problematic cases. Line 5 assigns a dimensionless quantity to a dimensionful array.  That may be OK but is ambiguous. Line 6 is especially problematic as we are now assigning completely invalid units to each other and the code is just happy to go along.

**Comment by ngoldbaum** [comment]
Line 5 is supposed to work that way. Line 6 should probably be an error. While fixing that, it would probably also be worth checking that if you tried to insert 3 meters, you’d end up with an entry that contains 300 centimeters.

**Comment by neutrinoceros** [comment]
I believe the first case working silently (`x[1] = 4`)  is the intended behaviour. It would also break copious amounts of code to change it.

In the second case however I may be convinced that adding some some input validation would be a good idea, and I don't _think_ it would break much dependent code.

I'll let other maintainers weigh in.

**Comment by neutrinoceros** [comment]
oh, nice race condition, @ngoldbaum !
Since there seems to be a consensus on the second case, and given the timing, let's discuss what version should such a change be aimed at. I think it qualifies well for version 3.0.0 since it's potentially breaking, WDYT ?

**Comment by ngoldbaum** [comment]
Yup, 3.0 will be the next major release, unless we back or disable all the `__array_function__` stuff to do a stopgap release, which I definitely don’t want to do.

**Comment by neutrinoceros** [comment]
What I meant was: given the two release branches we already have (2.9.x and treating main as 3.0.x), where should this land ? But the conclusion remains unchanged 😅 

Anyway, we'd accept a PR adding input validation for arrays element assignments !

**Comment by ngoldbaum** [comment]
Oh I see what you're saying.

Well, given that this happens:

```
In [1]: import unyt as u

In [2]: a = [1, 2, 3]*u.cm

In [3]: a[1] = 2*u.m

In [4]: a
Out[4]: unyt_array([1, 2, 3], 'cm')
```

it probably could be considered a bugfix, and any new errors would be coming from buggy code. I guess it could break code that depends on the buggy behavior above. Not really sure if there's a good way to detect that.

**Comment by neutrinoceros** [comment]
good call, given your new example I agree we could treat it as a bugfix.
@MatthieuSchaller, do you have any interest in doing the work ? Otherwise I'm happy to step in but I may not for a couple days.

**Comment by ngoldbaum** [comment]
I'm actually poking at it right now...

**Comment by MatthieuSchaller** [comment]
> good call, given your new example I agree we could treat it as a bugfix. @MatthieuSchaller, do you have any interest in doing the work ? Otherwise I'm happy to step in but I may not for a couple days.

I would if I had the time and the actual python knowledge to do it cleanly. 

Happy to test and complain (also positively!) if needed though.

**Comment by neutrinoceros** [comment]
@MatthieuSchaller feel free to test the patch #375 !

**Comment by neutrinoceros** [comment]
This is fixed in unyt 2.9.5
